### DISCUSSION TITLE
Custom script for new IP - ObjectVar

### DISCUSSION THREAD

**Discussion by jahknem** [netbox-community/netbox#9326] [Q&A]
Hi!

I'm currently writing a script which is supposed to add a new router, connect it to the respective edge router and populate the relevant interfaces with available IP adresses. It is supposed to get these IP adresses from prefixes with a specific role.

What I'm struggling with is how to populate an objectVar() with IP addresses and prefixes which do not exist yet in the database. I understand that the prefix.get_available_ips() and prefix.get_available_prefixes() function will get me these nonexisting objects, but I'm missing how to get these into the options a user is presented with when using objectVar().

Here is what I've got up to now:

```
parentPrefix = ObjectVar(
        label = "Parent Prefix",
        description="Available Prefixes",
        model=Prefix,
        query_params={
            'role': 'customer-lans'
        }
    )
    
    childIP = ObjectVar(
        label = "Customer IP",
        description = "Choose an available Ip for the customer",
        model=IPAddress,
        query_params=parentPrefix.get_available_ips()
    )
```

The error it throws is "'ObjectVar' object has no attribute 'get_available_ips'". 

That makes in essence two questions:

1. How do I use functions on the result of an objectVar within another ObjectVar?
2. How do I get an ObjectVar to show me the available IP addresses or child prefixes of an existing prefix?

**Comment by DanSheps** [comment]
Query parameters should be a dictionary of parameters to pass to the API query.

Instead, what you likely need is to override the URL for this.  You likely need to write a custom field object for this use case.

**Reply by jahknem** [reply]
Hi Daniel, thank you for the reply!
I'm currently looking at custom fields documentation, but I feel like I'm missing some context here. I'm not trying to add a new attribute to a model. Instead I'm trying to simplify the planning process for adding a new device by automatically adding the respective cable connection, child prefixes and ip addresses.

Also, how do I override the URL, and which URL would I be overriding, the one from the query?

Thanks!

**Reply by DanSheps** [reply]
I am not talking about a custom field in netbox, I am referring to inheriting the custom field then overriding the URL parameter so that it doesn't resolve the URL for the IP Address model but instead resolves the URL for the available IP endpoints then provides you with the IP (and not the object ID)

**Reply by jahknem** [reply]
I'm terribly sorry, but I still don't think I understand, even though I'm looking at the source for the CustomField.
What I think you're telling me is that 
- either I need to use CustomField as the model parameter for ObjectVar.
- or I need to use IPAddress as the model parameter for ObjectVar

subsequently I need to override the URL the model queries. 

In both cases I do not see the parameter which I need to overwrite to change the URL. Could you point me to a file/issue/discussion where something similar has been done before?

**Reply by jeremystretch** [reply]
I'll try to clarify a bit: ObjectVar can only reference objects which already exist in the database, so what you're trying to do with it won't work. Dan suggested potentially creating your own custom *form* field type (similar to ObjectVar) to enable this functionality (not to be confused with "custom fields" for objects).

**Reply by jahknem** [reply]
Thank you, that makes sense. From looking at the source code of the existing form field types in netbox, I can see I need to learn a lot more about Django and Python in general. Thanks to both of you for your patience!

As I understand it now the URL which is meant is api_url in APISelect. I'll try to create a custom form field type and will either post my solution here or (more likely) come back with even more questions.

**Reply by jahknem** [reply]
More Questionmarks:
I've now created a class "AvailableIPObjectVar" which does exactly the same thing as objectVar but should overwrite the api_url. Obviously this is still far from perfect as I'm still struggling with setting the id that defines which prefix should be used.

Also it does not work.

Class which defines AvailableIPObjectVar
```
class AvailableIPObjectVar(ScriptVariable):
    form_field = DynamicModelChoiceField

    def __init__(self, model, my_api_url=None, query_params=None, null_option=None, *args, **kwargs):
        super().__init__(*args, **kwargs)

        if my_api_url is not None:
            self.form_field.widget(api_url=my_api_url)

        self.field_attrs.update({
            'queryset': model.objects.all(),
            'query_params': query_params,
            'null_option': null_option,
        })
```
Calling AvailableIPObjectVar within my existing script.
```
...
    childIPs = AvailableIPObjectVar(
        label = "Available IPs in Prefix",
        model=IPAddress,
        my_api_url="api/ipam/prefixes/10/available-ips/"
    )
...
```
AvailableIPObjectVar still gives me all existing IPs. I would have expected it to give me the available IPs of the prefix with the ID 10. 
To me it looks like either api_url gets overwritten at a later time or what I'm overriding is the wrong parameter. 

Question: What am I doing wrong, shouldn't this work?

**Comment by k1nky** [comment]
Hi,I made that component. It is based on ChoiceField and APISelect widget and supports selection of available models (such as VLAN, prefixes, addresses). The list of models is obtained dynamically via API. The component is provided as a [plugin](https://github.com/k1nky/netbox-scripthelper-plugin).

Perhaps this will be useful to someone else.
